common/src/chain/transaction/signed_transaction.rs

```diff
@@ -67,10 +113,64 @@ impl PartialEq for WithId<SignedTransaction> {
 
 impl Eq for WithId<SignedTransaction> {}
 
-// TODO(PR): enforce that inputs size is equal to signatures size when decoding
+// TODO enforce that inputs size is equal to signatures size when decoding
 
-// TODO(PR): add a check in check_transactions and ensure that all signed transactions have sizes equal in witness and inputs
-// TODO(PR): add tests to check that inputs and witnesses have the same size
-
-// TODO(PR) make the SignedTransaction serialization ignore the size of the witness vec and just use the size of the inputs
+// TODO make the SignedTransaction serialization ignore the size of the witness vec and just use the size of the inputs
 // NOTE: there might be difficulties there as Encode cannot fail. It may lead to accepting a panic there
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use crate::chain::TxInput;
+    use crate::primitives::H256;
+
+    #[test]
+    fn require_inputs_witnesses_same_size() {
+        let hash0 = H256([0x50; 32]);
+        let hash1 = H256([0x51; 32]);
+        let hash2 = H256([0x52; 32]);
+
+        let ins0: Vec<TxInput> = [TxInput::new(Id::<Transaction>::new(hash0).into(), 5)].to_vec();
+        let ins1: Vec<TxInput> = [
+            TxInput::new(Id::<Transaction>::new(hash1).into(), 3),
+            TxInput::new(Id::<Transaction>::new(hash2).into(), 0),
+        ]
+        .to_vec();
+
+        let tx = Transaction::new(0x00, vec![], vec![], 0x01).unwrap();
+        assert!(SignedTransaction::new(tx.clone(), vec![]).is_ok());
+        assert!(matches!(
+            SignedTransaction::new(tx, vec![InputWitness::NoSignature(None)]),
+            Err(TransactionCreationError::InvalidWitnessCount)
+        ));
```

Comment:
Could this be an `assert_eq!`? (A couple more of these patterns below)

Reply:
Nope, tried it but `Result` doesn't implement `Eq`.

Reply:
Is that because the `SignedTransaction` type is not `Eq`? Maybe it would make sense to add it.

Reply:
It was because the error type was not `Eq`. Added a derive for that and changed to `assert_eq` as you suggested.